### DISCUSSION TITLE
Commands from host computer

### DISCUSSION THREAD

**Discussion by bwhetsel** [mciantyre/teensy4-rs#161] [General]
Is there a good example for sending commands from a host computer to the teensy over the USB port? 

I appreciate all logging capability but I am looking for an example that allows me to send commands to the teensy over USB? 

Even just an echo back example would really help.

**Comment by mciantyre** [comment]
There's [an example](https://github.com/imxrt-rs/imxrt-hal/blob/main/examples/rtic_usb_serial.rs) of USB serial echo in the imxrt-hal project. That example runs on a Teensy 4. You'll need to configure and control your own USB device stack.

**Comment by bwhetsel** [comment]
Thanks for the quick response! I had already tried the example you posted and I couldn’t get it working on Windows, I know there are a couple issues with windows usb. Is there something I can change to get it working?

**Reply by mciantyre** [reply]
Here's three options I would try:

Reduce [the USB speed](https://github.com/imxrt-rs/imxrt-hal/blob/715a6434850d98fb13217abfc09bcaa0a09963d1/examples/rtic_usb_serial.rs#L41) from `High` to `LowFull`. My understanding of the issue is that the high-speed USB serial class doesn't correctly present itself to a Windows host. If you reduce the speed, then the serial port might work correctly on Windows. Of course, it'll work slower. This should be the easiest change to try.

Design directly to the [`CdcAcmClass`](https://docs.rs/usbd-serial/0.2.2/usbd_serial/struct.CdcAcmClass.html) instead of `SerialPort`. As long as the max packet size is 512 bytes, then Windows should be happy with the serial class. If you need an example of how to use `CdcAcmClass`, check out the [USB device backend for the logging package](https://github.com/imxrt-rs/imxrt-hal/blob/main/logging/src/usbd.rs).

Patch the `usbd-serial` package. For high-speed operation, the serial port should allocate [a max packet size](https://github.com/rust-embedded-community/usbd-serial/blob/7331c49d7a6e432e1db66d350152116ca4437f3e/src/serial_port.rs#L94) of 512 bytes. I think the [interrupt endpoint interval](https://github.com/rust-embedded-community/usbd-serial/blob/7331c49d7a6e432e1db66d350152116ca4437f3e/src/cdc_acm.rs#L79) also needs to change; [here](https://github.com/mciantyre/usbd-serial/commit/80dec6f4ff4d5908a6cbd015428031194f9fbe2d) are my old notes and test values. Note that hard-coding these new values will break low/full speed device operation.

**Comment by stappersg** [comment]
On Sat, Aug 10, 2024 at 05:55:25AM -0700, bwhetsel wrote:
>  ... couldn’t get it working on Windows, I know there are a couple
>  issues with windows usb. Is there something I can change to get
>  it working?

Run the experiment on Linux.
